src/Build/Graph/ProjectGraph.cs

```diff
@@ -303,17 +334,63 @@ public ProjectGraph(IEnumerable<ProjectGraphEntryPoint> entryPoints, ProjectColl
             return targetLists;
         }
 
+        /// <summary>
+        /// A callback used for constructing a <see cref="ProjectInstance"/> for a specific
+        /// <see cref="ProjectGraphEntryPoint"/> instance.
+        /// </summary>
+        /// <param name="projectPath">The path to the project file to parse.</param>
+        /// <param name="projectCollection">The <see cref="ProjectCollection"/> context for parsing.</param>
+        /// <returns>A <see cref="ProjectInstance"/> instance. This value must not be null.</returns>
+        /// <remarks>
+        /// The default version of this delegate used by ProjectGraph simply calls the
+        /// ProjectInstance constructor with information from the parameters. This delegate
+        /// is provided as a hook to allow scenarios like creating a <see cref="Project"/>
+        /// instance before converting it to a ProjectInstance for use by the ProjectGraph.
+        ///
+        /// The returned ProjectInstance will be stored and provided with the ProjectGraph.
+        /// If this callback chooses to generate an immutable ProjectInstance, e.g. by
+        /// using <see cref="Project.CreateProjectInstance()"/> with the flag
+        /// <see cref="ProjectInstanceSettings.Immutable"/>, the resulting ProjectGraph
+        /// nodes might not be buildable.
+        /// </remarks>
+        public delegate ProjectInstance ProjectInstanceFactoryFunc(
+            string projectPath,
+            Dictionary<string, string> globalProperties,
+            string toolsVersion,
```

Comment:
Tools version is going away, so we should not include it. It's not part of the public API at all for `ProjectGraph`

Reply:
Also is there a reason to define a delegate instead of using `Func<string, Dictionary<string, string>, ProjectInstance>`? It does seem kind of long.

Reply:
@AndyGerlicher I started with Func<> but the param documentation got long, so moved to a delegate declaration to expand param definitions. I recommend staying with this format as squashing it back is pretty ugly.

Reply:
@dfederm Set to null? I was strictly copying the original code which sets that value to                 configurationMetadata.ToolsVersion

Reply:
`ConfigurationMetadata.ToolsVersion` is hardcoded to `MSBuildConstants.CurrentToolsVersion` in this case though: https://github.com/Microsoft/msbuild/blob/exp/net472/src/Build/BackEnd/Shared/ConfigurationMetadata.cs#L49

Null probably would work too I'd think.